prog/dump/i2cget.8
.TH I2CGET 8 "June 2007"
.SH "NAME"
i2cget \- read from I2C/SMBus chip registers

.SH SYNOPSIS
.B i2cget
.RB [ -f ]
.RB [ -y ]
.I i2cbus
.I chip-address
.RI [ data-address " [" mode ]]
.br
.B i2cget
.B -V

.SH DESCRIPTION
i2cget is a small helper program to read registers visible through the I2C
bus (or SMBus).

.SH OPTIONS
.TP
.B -V
Display the version and exit.
.TP
.B -f
Force access to the device even if it is already busy. By default, i2cget
will refuse to access a device which is already under the control of a
kernel driver. Using this flag is dangerous, it can seriously confuse the
kernel driver in question. It can also cause i2cget to return an invalid
value. So use at your own risk and only if you know what you're doing.
.TP
.B -y
Disable interactive mode. By default, i2cget will wait for a confirmation
from the user before messing with the I2C bus. When this flag is used, it
will perform the operation directly. This is mainly meant to be used in
scripts. Use with caution.
.PP
There are two required options to i2cget. \fIi2cbus\fR indicates the number
of the I2C bus to be scanned.  This number should correspond to one of
the busses listed by \fIi2cdetect -l\fR. \fIchip-address\fR specifies the
address of the chip on that bus, and is an integer between 0x03 and 0x77.
.PP
\fIdata-address\fR specifies the address on that chip to read from, and is
an integer between 0x00 and 0xFF. If omitted, the currently active register
will be read (if that makes sense for the considered chip).
.PP
The \fImode\fR parameter, if specified, is one of the letters \fBb\fP,
\fBw\fP or \fBc\fP, corresponding to a read byte data, a read word data or a
write byte/read byte transaction, respectively. A \fBp\fP can also be appended
to the \fImode\fR parameter to enable PEC. If the \fImode\fR parameter is omitted,
i2cget defaults to a read byte data transaction, unless \fIdata-address\fR is
also omitted, in which case the default (and only valid) transaction is a
single read byte.

.SH WARNING
i2cget can be extremely dangerous if used improperly. I2C and SMBus are designed
in such a way that an SMBus read transaction can be seen as a write transaction by
certain chips. This is particularly true if setting \fImode\fR to \fBcp\fP (write byte/read
byte with PEC). Be extremely careful using this program.

.SH SEE ALSO
i2cdump(8), i2cset(8)

.SH AUTHOR
Jean Delvare
http://www.lm-sensors.org/
This manual page was strongly inspired from those written by David Z Maze
for i2cset.


CHANGES
lm_sensors CHANGES file
-----------------------

SVN HEAD
  File i2c-dev.h: Support I2C block reads with specified length
  File sensors.conf.eg: Add a dme1737 section
  File lm_sensors.sysconfig: Delete, now generated by sensors-detect
  Library: Clean up pwm symbol names translation
           Add dme1737 support
           Add applesmc support
           Add w83627ehf support (for Linux 2.4)
           Add support for the w83627ehf VID function
           Add support for the w83627ehf thermal sensor types
  Man page sensors.conf.5: Update the chip statement section
  Module i2c-nforce2: Add nForce MCP61, MCP65 support (2.6 backport)
  Module w83627ehf: New (2.6 backport)
  Programs i2cdump, i2cget, i2cset: Do not force the slave address by default
  Program i2cdump: Use the new I2C block read function
  Program isadump: Detect when address bit 7 is a busy flag
                   Fix Super-I/O exit sequence for Winbond/Fintek chips
  Program fancontrol: Use let for arithmetic evaluation
                      No longer need awk and grep
                      Limit calls to external programs (Linux 2.6 only)
                      Check for configuration file validity
                      Support optional min and max PWM values
                      Better integration with init scripts
  Program pwmconfig: Use smaller steps for low PWM values
                     Support optional min and max PWM values
  Program mkpatch: Include w83627ehf
  Program rrd: Support non-i2c devices (hwmon class)
               Add missing shell declarations
               Update URI
  Program sensors: Drop ddcmon and eeprom support
                   Add dme1737 support
                   Add applesmc support
                   Add max6680 support
                   Fix alignment of alarm for one-limit temperatures
                   Add w83627ehf support (for Linux 2.4)
                   Add support for the w83627ehf VID function
                   Add support for the w83627ehf thermal sensor types
  Program sensors-detect: Stop Super-I/O probe after first family success
                          Fix SMSC DME1737 detection
                          Add /usr/sbin to the PATH (#2199)
                          Add ATI SB700 detection
                          Add Maxim MAX6680/MAX6681 detection
                          Add IT8726F detection
                          Fix IPMI support for 2.6 kernels
                          Add detection for non-standard SMSC Super-I/Os
                          No longer depend on i2cdetect
                          Add SMSC SCH5317 detection
                          Drop detection of most non-sensors I2C chips


2.10.3 (20070319)
  Library: Do not skip i2c adapters with only a class device and no
            physical device. This change is required to survive the
            planned struct class_dev removal from future 2.6 kernels.
           Add support for the SMSC LPC47M292 fans (prefix smsc47m2)
           Recognize i2c-isa-based devices as such even after class_dev
	    is gone
           Compile with -D_REENTRANT
           Fix support of Linux 2.6's max6650 driver
           Fix max1619 support
  Man page i2cdetect.8: Scanning range can be restricted
  Module bmcsensors: Fix debugging messages
  Module i2c-ali1563: Improve the status messages (2.6 backport)
                      Fix device initialization (2.6 backport)
  Module i2c-amd8111: Cleanups (2.6 backport)
                      Fix initialization race (2.6 backport)
  Module i2c-nforce2: Drop unused reference to pci_dev (2.6 backport)
  Module i2c-piix4: Add ATI SB600 support (2.6 backport)
  Module i2c-viapro: Add CX700 support (2.6 backport)
  Module smsc47m1: Add SMSC LPC47M292 support
                   Get rid of a useless semaphore (2.6 backport)
  Programs i2cdetect, i2cdump, isadump: Flush output in real time
  Program i2cdump: Cleaner output in SMBus block read mode
  Program fancontrol.pl: Fix pwmN_enable access (#2190)
  Program sensors: Add smsc47m2 support (fans)
                   Fix -A option (hide adapter name) on error
  Program sensors-detect: Only probe relevant I2C addresses
                          Drop ARP-capable device detection
                          Add Maxim MAX6655/MAX6656 detection
                           (Olexiy Avramchenko)
                          Add Maxim MAX1668, MAX1805, MAX1989 detection
                          Add VIA CX700 detection
                          Drop legacy W83627HF ISA detection
                          Don't bufferize stdout during probes
                          Speed up the LM75 and LM77 detection
                          Clean up the EEPROM detection
                          Add Dallas DS75 detection
                          Add SMSC LPC47M292 detection
                          Add nVidia MCP61 and MCP65 detection
                          Fix ADM1028 detection
                          Add Nat. Semi. PC8347L Super-I/O detection
                          Add SMSC SCH5504 Super-I/O detection (no sensors)
                          Make the LM78 and W83781D detection safer


2.10.2 (20070115)
  RPM: Obsolete, removed
  File doc/fan-divisors: Update
  File sensors_vid.h: Trim VID values to correct number of bits (2.6 backport)
  Library: Add support for the pc87247 driver (fans only)
           Probe for busses before chips
           Drop support for algorithm names
           Add support for coretemp driver   
           Major update/optimization of config-file scanner
           Add config-file scanner regression tests
           Add f71872f support
           Add w83627dhg support (David Holl, #2157)
           Get the i2c adapter names from classdev.name
           Add adm1029 support (Corentin Labbe)
  Man page i2cdetect.8: Describe the output convention
  Man page sensors.1: Update (option -c) and clean up
  Module bmcsensors: Fix several warnings
                     Fix fragile structure initialization

prog/dump/i2cget.c
/*
    i2cget.c - A user-space program to read an I2C register.
    Copyright (C) 2005       Jean Delvare <khali@linux-fr.org>

    Based on i2cset.c:
    Copyright (C) 2001-2003  Frodo Looijaard <frodol@dds.nl>, and
                             Mark D. Studebaker <mdsxyz123@yahoo.com>
    Copyright (C) 2004-2005  Jean Delvare <khali@linux-fr.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "util.h"
#include "i2cbusses.h"
#include "i2c-dev.h"
#include "version.h"

void help(void) __attribute__ ((noreturn));

void help(void)
{
	fprintf(stderr, "Syntax: i2cget [-f] [-y] I2CBUS CHIP-ADDRESS "
	        "[DATA-ADDRESS [MODE]]\n"
	        "        i2cget -V\n"
	        "  MODE can be: 'b' (read byte data, default)\n"
	        "               'w' (read word data)\n"
	        "               'c' (write byte/read byte)\n"
		"  If DATA-ADDRESS is omitted, a single read byte command is "
		"issued\n"
	        "  Append 'p' to MODE for PEC checking\n"
	        "  I2CBUS is an integer\n");
	print_i2c_busses(0);
	exit(1);
}

int check_funcs(int file, int i2cbus, int size, int daddress, int pec)
{
	unsigned long funcs;

	/* check adapter functionality */
	if (ioctl(file, I2C_FUNCS, &funcs) < 0) {
		fprintf(stderr, "Error: Could not get the adapter "
		        "functionality matrix: %s\n", strerror(errno));
		return -1;
	}

	switch (size) {
	case I2C_SMBUS_BYTE:
		if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE)) {
			fprintf(stderr, "Error: Adapter for i2c bus %d does "
			        "not have read byte capability\n", i2cbus);
			return -1;
		}
		if (daddress >= 0
		 && !(funcs & I2C_FUNC_SMBUS_WRITE_BYTE)) {
			fprintf(stderr, "Error: Adapter for i2c bus %d does "
			        "not have write byte capability\n", i2cbus);
			return -1;
		}
		break;

	case I2C_SMBUS_BYTE_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE_DATA)) {
			fprintf(stderr, "Error: Adapter for i2c bus %d does "
			        "not have read byte data capability\n", i2cbus);
			return -1;
		}
		break;

	case I2C_SMBUS_WORD_DATA:
		if (!(funcs & I2C_FUNC_SMBUS_READ_WORD_DATA)) {
			fprintf(stderr, "Error: Adapter for i2c bus %d does "
			        "not have read word data capability\n", i2cbus);
			return -1;
		}
		break;
	}

	if (pec
	 && !(funcs & (I2C_FUNC_SMBUS_HWPEC_CALC | I2C_FUNC_I2C))) {
		fprintf(stderr, "Warning: Adapter for i2c bus %d does "
		        "not seem to support PEC\n", i2cbus);
	}

	return 0;
}

int confirm(const char *filename, int address, int size, int daddress, int pec)
{
	int dont = 0;

	fprintf(stderr, "WARNING! This program can confuse your I2C "
		"bus, cause data loss and worse!\n");

	/* Don't let the user break his/her EEPROMs */
	if (address >= 0x50 && address <= 0x57 && pec) {
		fprintf(stderr, "STOP! EEPROMs are I2C devices, not "
			"SMBus devices. Using PEC\non I2C devices may "
			"result in unexpected results, such as\n"
			"trashing the contents of EEPROMs. We can't "
			"let you do that, sorry.\n");
		return 0;
	}

	if (size == I2C_SMBUS_BYTE && daddress >= 0 && pec) {
		fprintf(stderr, "WARNING! All I2C chips and some SMBus chips "
		        "will interpret a write\nbyte command with PEC as a"
		        "write byte data command, effectively writing a\n"
		        "value into a register!\n");
		dont++;
	}

	fprintf(stderr, "I will read from device file %s, chip "
		"address 0x%02x, ", filename, address);
	if (daddress < 0)
		fprintf(stderr, "current data\naddress");
	else
		fprintf(stderr, "data address\n0x%02x", daddress);
	fprintf(stderr, ", using %s.\n",
		size == I2C_SMBUS_BYTE ? (daddress < 0 ?
		"read byte" : "write byte/read byte") :
		size == I2C_SMBUS_BYTE_DATA ? "read byte data" :
		"read word data");
	if (pec)
		fprintf(stderr, "PEC checking enabled.\n");

	fprintf(stderr, "Continue? [%s] ", dont ? "y/N" : "Y/n");
	fflush(stderr);
	if (!user_ack(!dont)) {
		fprintf(stderr, "Aborting on user request.\n");
		return 0;
	}

	return 1;
}

int main(int argc, char *argv[])
{
	char *end;
	int res, i2cbus, address, file;
	int size = I2C_SMBUS_BYTE_DATA;
	int daddress;
	char filename[20];
	int pec = 0;
	int flags = 0;
	int force = 0, yes = 0, version = 0;

	/* handle (optional) flags first */
	while (1+flags < argc && argv[1+flags][0] == '-') {
		switch (argv[1+flags][1]) {
		case 'V': version = 1; break;
		case 'f': force = 1; break;
		case 'y': yes = 1; break;
		default:
			fprintf(stderr, "Warning: Unsupported flag "
				"\"-%c\"!\n", argv[1+flags][1]);
			help();
			exit(1);
		}
		flags++;
	}

	if (version) {
		fprintf(stderr, "i2cget version %s\n", LM_VERSION);
		exit(0);
	}

	if (argc - flags < 3)
		help();

	i2cbus = strtol(argv[flags+1], &end, 0);
	if (*end || i2cbus < 0 || i2cbus > 0x3f) {
		fprintf(stderr, "Error: I2CBUS argument invalid!\n");
		help();
	}

	address = strtol(argv[flags+2], &end, 0);
	if (*end || address < 3 || address > 0x77) {
		fprintf(stderr, "Error: Chip address invalid!\n");
		help();
	}

	if (!(flags+3 < argc)) {
		size = I2C_SMBUS_BYTE;
		daddress = -1;
	} else {
		daddress = strtol(argv[flags+3], &end, 0);
		if (*end || daddress < 0 || daddress > 0xff) {
			fprintf(stderr, "Error: Data address invalid!\n");
			help();
		}
	}

	if (flags+4 < argc) {
		switch (argv[flags+4][0]) {
		case 'b': size = I2C_SMBUS_BYTE_DATA; break;
		case 'w': size = I2C_SMBUS_WORD_DATA; break;
		case 'c': size = I2C_SMBUS_BYTE; break;
		default:
			fprintf(stderr, "Error: Invalid mode!\n");
			help();
		}
		pec = argv[flags+4][1] == 'p';
	}

	file = open_i2c_dev(i2cbus, filename, 0);
	if (file < 0
	 || check_funcs(file, i2cbus, size, daddress, pec)
	 || set_slave_addr(file, address, force))
		exit(1);

	if (!yes && !confirm(filename, address, size, daddress, pec))
	 	exit(0);

	if (pec && ioctl(file, I2C_PEC, 1) < 0) {
		fprintf(stderr, "Error: Could not set PEC: %s\n",
		        strerror(errno));
		exit(1);
	}

	switch (size) {
	case I2C_SMBUS_BYTE:
		if (daddress >= 0) {
			res = i2c_smbus_write_byte(file, daddress);
			if (res < 0)
				fprintf(stderr, "Warning - write failed\n");
		}
		res = i2c_smbus_read_byte(file);
		break;
	case I2C_SMBUS_WORD_DATA:
		res = i2c_smbus_read_word_data(file, daddress);
		break;
	default: /* I2C_SMBUS_BYTE_DATA */
		res = i2c_smbus_read_byte_data(file, daddress);
	}
	close(file);

	if (res < 0) {
		fprintf(stderr, "Error: Read failed\n");
		exit(2);
	}

	printf("0x%0*x\n", size == I2C_SMBUS_WORD_DATA ? 4 : 2, res);

	exit(0);
}
